File: app/src/main/java/karrel/com/bluetoothmanager/presenter/MainPresenterImpl.java
package karrel.com.bluetoothmanager.presenter;

import android.bluetooth.BluetoothDevice;
import android.content.Context;

import com.karrel.mylibrary.RLog;

import karrel.com.btconnector.btmanager.BluetoothListener;
import karrel.com.btconnector.btmanager.BluetoothManager;
import karrel.com.btconnector.btmanager.BluetoothManagerable;
import karrel.com.btconnector.btscanner.DiscoveryScanner;

/**
 * Created by Rell on 2018. 3. 23..
 */

public class MainPresenterImpl implements MainPresenter {
    private MainPresenter.View view;
    private BluetoothManager bluetoothManager;
    // 블루투스 기기
    private BluetoothDevice bluetoothDevice;

    public MainPresenterImpl(MainPresenter.View view, Context context) {
        this.view = view;

        bluetoothManager = BluetoothManager.getInstance(context);
        bluetoothManager.setScanner(BluetoothManager.Scanner.DISCOVERY);
        bluetoothManager.addBluetoothCallback(bluetoothListener);
    }

    @Override
    public void searchBluetoothDevices() {
        RLog.e();
        bluetoothManager.startBluetoothDeviceScan();
    }

    // BT 설정 성공
    @Override
    public void enabledBluetooth() {
        // 블루투스 탐색 시작
        searchBluetoothDevices();
    }

    @Override
    public void dissableBluetooth() {
        // BT 설정 실패
        RLog.e();
    }

    @Override
    public void connectBluetooth() {
        RLog.e();
        if (bluetoothDevice == null) return;
        bluetoothManager.connect(bluetoothDevice);
    }

    @Override
    public void disconnectBluetooth() {
        bluetoothManager.disConnect();
    }

    private final BluetoothListener bluetoothListener = new BluetoothListener() {
        @Override
        public void onConnectedSuccess(String deviceName) {
            RLog.e("deviceName : " + deviceName);
        }

        @Override
        public void onConnecting(String deviceName) {
            RLog.e("deviceName : " + deviceName);
        }

        @Override
        public void onInitalized(String deviceName) {
            RLog.e("deviceName : " + deviceName);
        }

        @Override
        public void onMessageSend(byte[] writeBuf) {
            RLog.e("writeBuf : " + writeBuf);
        }

        @Override
        public void onReadMessage(byte[] readBuf) {
            RLog.e("readBuf : " + readBuf);
        }

        @Override
        public void onStartConnect(String deviceName) {
            RLog.e("deviceName : " + deviceName);
        }

        @Override
        public void onConnectedFail(String name) {
            RLog.e("deviceName : " + name);
        }

        @Override
        public void deniedPermission() {
            RLog.e();
        }

        @Override
        public void requireEnableBt() {
            RLog.e();
            view.requireEnableBt();
        }

        @Override
        public void onSearchedDevice(BluetoothDevice device) {
            RLog.d(String.format("device : %s, address : %s", device.getName(), device.getAddress()));

            String name = "Phantom";
            String address = "5C:B6:CC:04:15:66";
//            String address = "4C:E1:73:B6:A8:1A"; // 아닌듯
//            String address = "5C:B6:CC:04:18:1C"; // 아닌듯
//            String address = "5C:B6:CC:03:A6:CC"; // 아닌듯
            boolean isCollect = device.getName().equals(name) && device.getAddress().equals(address);
            if (isCollect) {
                bluetoothDevice = device;
            }
        }
    };
}

File: app/src/main/java/karrel/com/bluetoothmanager/presenter/MainPresenterImpl.kt
package karrel.com.bluetoothmanager.presenter

import android.Manifest
import android.bluetooth.BluetoothDevice
import android.content.Context
import android.support.annotation.RequiresPermission
import android.support.v4.content.ContextCompat
import karrel.com.bluetoothmanager.util.ByteConverter
import karrel.com.btconnector.btmanager.BluetoothListener
import karrel.com.btconnector.btmanager.BluetoothManager

/**
 * Created by Rell on 2018. 3. 23..
 */

class MainPresenterImpl(private val view: MainPresenter.View, context: Context) : MainPresenter {

    private val bluetoothManager: BluetoothManager = BluetoothManager.getInstance(context)
    // 블루투스 기기

    private val bluetoothListener = object : BluetoothListener {
        override fun onFailedConnect(deviceName: String?) {
            println("onFailedConnect($deviceName)")
        }

        override fun onLostedConnect(deviceName: String?) {
            println("onLostedConnect($deviceName)")
        }

        override fun onConnected(deviceName: String?) {
            println("onConnected($deviceName)")
        }

        override fun onConnecting(deviceName: String?) {
            println("onConnecting($deviceName)")
        }

        override fun onListening(deviceName: String?) {
            println("onListening($deviceName)")
        }

        override fun onSendMessage(writeBuf: ByteArray?) {
            println("onSendMessage()")
        }

        override fun onReadMessage(readBuf: ByteArray?) {
            println("onReadMessage(${ByteConverter.byteToHex(readBuf)})")
        }

        override fun onStartConnect(deviceName: String?) {
            println("onStartConnect($deviceName)")
        }

        override fun onDoingNothing(deviceName: String?) {
            println("onDoingNothing($deviceName)")
        }

        override fun deniedPermission() {
            println("deniedPermission()")
        }

        override fun requireEnableBt() {
            println("requireEnableBt()")
            view.requireEnableBt()
        }

        override fun onSearchedDevice(device: BluetoothDevice?) {

            val text = "device : ${device?.name}, address : ${device?.address}, bondState : ${device?.bondState}"
            println("onSearchedDevice : $text")

            view.addSearchedDevice(device)

        }
    }

    init {

        bluetoothManager.setScanner(BluetoothManager.Scanner.DISCOVERY)
        bluetoothManager.addBluetoothCallback(bluetoothListener)
    }


    override fun onResume() {
    }

    override fun onStop() {
        bluetoothManager.stopBluetoothDeviceScan()
    }

    override fun onDestroy() {
    }

    @RequiresPermission(allOf = [Manifest.permission.BLUETOOTH, Manifest.permission.BLUETOOTH_ADMIN, Manifest.permission.ACCESS_COARSE_LOCATION])
    override fun searchBluetoothDevices() {
        bluetoothManager.startBluetoothDeviceScan()
        view.clearLog()
    }

    // BT 설정 성공
    @RequiresPermission(allOf = [Manifest.permission.BLUETOOTH, Manifest.permission.BLUETOOTH_ADMIN, Manifest.permission.ACCESS_COARSE_LOCATION])
    override fun enabledBluetooth() {
        // 블루투스 탐색 시작
        searchBluetoothDevices()
    }

    override fun dissableBluetooth() {
        // BT 설정 실패
    }

    override fun connectBluetooth(data: BluetoothDevice) {
        bluetoothManager.connect(data)
    }

    override fun disconnectBluetooth() {
        bluetoothManager.disConnect()
    }
}
